Python bindings

bunny-therapist opened the pull request:
Open issues:

**RuntimeError** (Edit: Fixed)
The concurrency test has some warning:
```
PytestUnhandledThreadExceptionWarning: Exception in thread Thread-4 (get_n_best)

  Traceback (most recent call last):
    File "<...>threading.py", line 1052, in _bootstrap_inner
      self.run()
    File "<...>threading.py", line 989, in run
      self._target(*self._args, **self._kwargs)
  RuntimeError: Already borrowed
```
I suppose it is even possible that right now the test is only passing because the threads just die immediately and thus "finish quickly". I should change the test to check for exceptions in the threads. This issue should then be investigated. It is imperative that we have concurrency in my opinion. Shouldn't multiple threads be able to each have an immutable reference to the same data and do whatever? Are we missing some send/sync stuff?

**Typing** (Edit: Fixed)
`Yake.__init__` takes either `language` or `stopwords`, but not both, and it must have one of them. I tried to represent this in the .pyi file with `typing.overload` but did not succeed. It would be nice if we could get this working.

**Building and environments**
Maturin can build sdists and wheels - I think - I would like to make sure the sdist is fine, and I would like to look into setting up some github workflows for building wheels for different systems.

**Readme and license** (Edit: Fixed)
I can reference the outer `README.md` and `LICENSE` files with a relative path in `python/Cargo.toml`, but then tox won't work because it won't find those files in the isolated build environment (`MANIFEST.in` cannot fix this because it cannot specify paths in its own parent directory). I'd rather have those in place and just drop tox (because we ship the package, not the tests/linting), but I am concerned about sdist distributions maybe having some issue with this as well, so I would at least like to test sdist stuff more.

**Version** (Edit: Resolved)
Do we want the python package to have the same version as the rust crate? If so, we would have to keep `version` in `python/Cargo.toml` aligned with `yake_rust/Cargo.toml`. We could of course use separate versions for them. If we want the same, it would be nice if we did not have to do that manually.

**Uploading** (Edit: Fixed)
I have not done anything with regards to uploading anything to pypi. I was thinking I would set up permissions for this project once the building is completely in place.

bunny-therapist commented:
Close #24

bunny-therapist commented:
Ok, I fixed the concurrency test (so that it is now failing).

bunny-therapist commented:
Alright. Fixed the concurrency issue. I just needed to make the self reference immutable.

bunny-therapist commented:
I suppose the python package should have its own README with python code examples.

xamgore commented:
Fantastic work 👀

Could you extend CONTRIBUTING.md? In case anyone opens an issue, we might ask him to propose a PR as well.

I have looked through the code, looks fine. I would switch off a few clippy rules.

- Does pyo3 copies data when passing it through?
- Curious, why yake-rust is not so fast on small texts and where is the problem
- Is it okay if I replace std::collections::HashMap with FxHashMap or AHashMap, which are not cryptographically safe and therefore faster?
- I am in the middle of rewriting the README, is it ok if we postpone the release till next Friday?

bunny-therapist commented:
> Fantastic work 👀
> 
> Could you extend CONTRIBUTING.md? In case anyone opens an issue, we might ask him to propose a PR as well.
>

Sorry, I don't quite understand what you are asking me to add.

> I have looked through the code, looks fine. I would switch off a few clippy rules.
> 
>     * Does pyo3 copies data when passing it through?
>

My understanding is that it converts python objects to rust objects and vice versa, based on some rules. So `str` becomes `String` going in and the returned `Vec` becomes a `list`. Not sure exactly what it means to "copy data" here. The original python object still exists in python, but you now have a converted rust object as well, so presumably "yes"?

>     * Curious, why yake-rust is not so fast on small texts and where is the problem
> 

I was a bit surprised by this too, because when I was first writing python bindings for yake-rust (before you and I got involved in the crate), this was not the case. I suspect we changed something that made it slower. Did you compare segtok with unicode-segmentation performance?

Of course, going between rust and python will incur some performance cost. There are some options one can try to improve it; but I don't want to do something I don't understand. Adding "frozen" to the Yake pyclass actually improved performance significantly.

>     * Is it okay if I replace std::collections::HashMap with FxHashMap or AHashMap, which are not cryptographically safe and therefore faster?
> 

You mean for stopwords and punctuation? As long as it is possible to convert a regular hashmap into these I think it is fine. Stopwords and punctuation will come into the rust code from python as `HashSet`, so I would have to convert it before putting it in the config. However, that is only done at class instantiation anyway (I actually decided to go with a class for this very purpose - I wanted to make sure we could reuse rust objects like stopwords etc without passing them back and forth between python and rust).

>     * I am in the middle of rewriting the README, is it ok if we postpone the release till next Friday?
> 

Sure.

> 
> I'll read the PR once again, as this time I've done it with my phone.

It is still just a draft. I need to understand the actual building process better and look into some stuff. Though we can of course merge this before everything is 100% done. I will convert it from a draft to a real PR when I feel good about it.

bunny-therapist commented:
I think we will just go with having different versioning for the python package and the rust crate. Could be useful, and I can't think of a way to unify them that isn't ridiculous.

bunny-therapist commented:
I don't quite understand this or if this is something that we should do or not: https://pyo3.rs/v0.23.4/performance.html#disable-the-global-reference-pool

bunny-therapist commented:
Alright, I think I need to go thru this: https://www.maturin.rs/distribution

xamgore commented:
Well, replacing `segtok` with just `.split(".").split_whitespace()` gives a 1000x speed up. When running a test with a small text (180 bytes) the time is mostly spent on initializing lazy regex automatons, but further tests run faster: 6ms vs. 100μs. In my benchmarks I see the slowest time, the median, the mean, the fastest.

bunny-therapist commented:
Interesting. Do you know if this also applies when calling from Python? I would assume so, but I will check.

bunny-therapist commented:
I don't actually see that kind of difference when running from python. For the short text, the time is
```
0.0001104000
0.0001050000
0.0001055000
0.0001173000
0.0001231000
0.0000935000
0.0001338000
0.0001003999
0.0001002001
0.0000936999
0.0000965000
```
Though I guess they are all like 100 microseconds like you say.

bunny-therapist commented:
For comparion, this is LIAAD/yake for the same short text:
```
0.0004106000
0.0003625000
0.0002954000
0.0004134000
0.0005442001
0.0003265999
0.0002978999
0.0003141001
0.0002775000
0.0003502000
0.0002877000
```

bunny-therapist commented:
Looking at the above figures, we _are_ faster than LIAAD/yake. The reason the tests are flaky and we do not always beat LIAAD/yake might be due to outliers. Here is a failing test:

the assertion is `assert dt_rust < dt_liaad`
and it evaluates to
```
assert 0.010449799941852689 < 0.0009973000269383192
```
So in this case, yake-rust was like a 100 times slower.

bunny-therapist commented:
Oh, ok. I think the "warm up" absolutely matters. I removed the "flaky" decorator and just ran the test for the short text, and it failed to beat LIAAD/yake every time. I added an extra run of yake-rust at the start of the test, discarding the result (just to have called it once already), and now yake-rust beats LIAAD/yake consistently.

I think I was just confused by "flaky" and the order of tests or something. I think the first time we call `get_n_best` it is slow, then faster subsequent times, beating LIAAD/yake consistently.

So it really _is_ a "warm-up" issue.

Would it be worth it to deal with this somehow? Or is it just acceptable?

bunny-therapist commented:
Does it help segtok that regexs are lazy?

xamgore commented:
> Does it help segtok that regexs are lazy?

I'll come back later to answer this.

For both benchmarks and latency-critical applications, I can expose a function like `warm_up` or `init`.

bunny-therapist commented:
I have added a github workflow triggered on "release" (so creating release via GitHub interface) which will build and upload to test.pypi (I will replace with the real pypi once it is confirmed to work).

The python wheels are platform dependent, so they need to be built for multiple platforms. I am also building for every supported python version. That could be dropped if we opt-in to the "abi3" interface, but it will allegedly come with an unknown performance cost. (I also have not really figured out how to do it properly.)

bunny-therapist commented:
I added CI that runs the python tests on pushes and PRs to master.

Hopefully we won't be using up all of Kyle's free github runner minutes. Apparently there are 2000 minutes per year, and the CI seems to consume like 5, so 400 runs, then the release jobs is probably more than that.

Should be fine, but hopefully we won't push just one commit at a time. We could also adjust this. It's just that if I am going to publish from GitHub master branch, there should be checks that verify that it is working. We also have no git hooks that check this.

bunny-therapist commented:
Note that the python jobs must necessarily build the rust crate also. I suppose we could have e.g. an initial CI job that runs "cargo check" or something that the other jobs depend on, so those jobs won't run if the code is fubar (of course, git hooks make this less likely).

xamgore commented on python/Cargo.toml:
```diff
@@ -0,0 +1,16 @@
+[package]
+name = "yake-rust-python-bindings"
+version = "1.0.0"
```
Do we keep it in sync with `yake_rust`?

bunny-therapist replied:
Not sure. I was thinking that we should, but I can't think of any way of doing that automatically.

Actually, maybe we should have it be different. Let's say the rust crate is super stable (in the future), but a new python version is released and one is deprecated. Then the python package should have a new release with support for the new version, but the rust crate does not need to be touched.

xamgore replied:
Ok, anyway. Must I publish `yake-rust-python-bindings` to crates.io?

bunny-therapist replied:
No. It should not be published.

bunny-therapist replied:
That name only exists because Cargo.toml needs a name. The name of the package is defined in pyproject.toml.

bunny-therapist replied:
I could have defined the name in Cargo.toml as well, but then I would have to (I assume) pick a different name because "yake-rust" is already taken. Doing it like this, the python package can be called "yake-rust".

xamgore replied:
The name is fine. Ok, I'll check the code once again locally in a few days

xamgore commented on python/src/lib.rs:
```diff
@@ -0,0 +1,77 @@
+    ///
+    /// Get the n best keywords from text.
+    #[pyo3(signature = (text, *, n))]
+    pub fn get_n_best(&self, py: Python, text: String, n: usize) -> Vec<(String, f64)> {
```
Is it `text: String` or a reference? `String` is owned, thus it is copied be thrown away immediately (as `segtok` uses `&str` and creates `Vec<String>`). No doubt `memcpy` is fast, but still

bunny-therapist replied:
It looks like maybe both are possible? Not sure. https://pyo3.rs/v0.23.4/conversions/tables.html

I will try with `&str`.

bunny-therapist replied:
Both work. I will change it to `&str`.

xamgore commented:
> Apparently there are 2000 minutes per year

We may ask Github team for more time.

bunny-therapist commented:
@xamgore When you publish yake-rust crate, do you do that from the yake_rust directory? Is that how it works? I was just concerned about the project root Cargo.toml containing "python" as a member when we don't want to publish what is in there. (I have no experience with publishing crates.)

xamgore commented:
https://doc.rust-lang.org/cargo/reference/manifest.html#the-publish-field

bunny-therapist commented:
Ah, so I don't need the version field in Cargo.toml. I will replace it with publish = false then.

bunny-therapist commented:
Do you think we could merge this this weekend?

bunny-therapist commented:
I locally built a wheel and I am currently testing yake-rust in a real usecase at work (another team was using yake - that is how I learned about it - I just made a version of their product but with yake-rust instead and run it alongside).

So far it seems superior. Much lower resource usage and much faster than LIAAD/yake. The only concern I have is that memory usage seems to trend upwards monotonically. I will see what is going on later in the week, but could it be that there is some sort of memory leak? (It is not clear to me how memory ownership and dropping etc is handled between rust and python. I am pretty sure nothing is going wrong natively in rust.)

xamgore commented:
Maybe there is something in the code to reference the binding's frozen class instances, which prevents GC collection. Or initializes multiple instances of yake per each thread. 

> Do you think we could merge this this weekend?

Sorry, was really busy.

bunny-therapist commented:
We'll see if the memory keeps ticking up.

The thing in question works like this: a python object has an attribute containing an instance of `yake_rust.Yake`. It is called through an async REST API, each request starts a new thread; the thread has access to the `yake_rust.Yake` instance (so it is a shared object). Each thread runs `get_n_best` and returns the result to the client, then the thread is finished.

There is just one `yake_rust.Yake` instance (at least there should be - it is only initialized in one place), so I can't imagine how there would be a leak.

bunny-therapist commented:
It ticks up, albeit slowly, but it does that for the regular version of the thing as well ("regular"=no yake-rust) so I suspect there might be a memory leak in another library and that we are innocent. Let's assume so until we learn otherwise.

xamgore commented:
Ok, I'm ready then. I will append formulas later, as I need them to be converted into svg which is not easy.

xamgore reviewed (approved):
Everything looks fine. I wish there was a section in `CONTRIBUTING.md` file to explain how to even build the project and check it. Like, I am out of python ecosystem, I know there is `uv` but have no idea how to start it running fast. There is already rust section in there.

bunny-therapist commented:
Sure, I can add that.

xamgore commented:
I have rebased my commits onto yours branch to make the history more linear.

xamgore commented:
Do you wanna bring another 7-10% speedup for free?

before:
```
Timer precision: 41 ns
bench          fastest       │ slowest       │ median        │ mean          │ samples │ iters
├─ text_3kb    3.978 ms      │ 13.9 ms       │ 4.048 ms      │ 4.077 ms      │ 2443    │ 2443
├─ text_100kb  139 ms        │ 145.6 ms      │ 140.2 ms      │ 140.5 ms      │ 100     │ 100
╰─ text_170b   187.8 µs      │ 401.5 µs      │ 191.7 µs      │ 193.4 µs      │ 51673   │ 51673
```

after:
```
Timer precision: 41 ns
bench          fastest       │ slowest       │ median        │ mean          │ samples │ iters
├─ text_3kb    3.808 ms      │ 13.95 ms      │ 3.86 ms       │ 3.879 ms      │ 2568    │ 2568
├─ text_100kb  129.9 ms      │ 136.1 ms      │ 131.2 ms      │ 131.5 ms      │ 100     │ 100
╰─ text_170b   175.4 µs      │ 370.3 µs      │ 178.4 µs      │ 179.8 µs      │ 55580   │ 55580
```
